compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -975,12 +975,15 @@ fn variant_info_for_generator<'tcx>(
             if variant_size == Size::ZERO {
                 variant_size = upvars_size;
             }
-            // We need to add the discriminant size back into min_size, since it is subtracted
-            // later during printing.
-            variant_size += match tag_encoding {
-                TagEncoding::Direct => tag.size(cx),
-                _ => Size::ZERO,
-            };
+
+            if layout.fields.offset(tag_field) >= variant_size {
```

Comment:
Nit: It took me a while to understand why this change was correct, probably mostly because the `variant_size` variable is a little confusingly named. Maybe something like `last_field_end` would better?

Thinking about this a little more, what this code really wants to do is `variant_size.max(tag_end)`, maybe that would also make it simpler.

Reply:
This is true, all of this code is trying to fit generator layout into a normal layout shaped hole and it's not going great. The current name is fine too and yeah, spending too much time on this is probably not worth it.

Reply:
But a quick comment would be nice.